src/Symfony/Component/Filesystem/Filesystem.php

```diff
@@ -453,6 +453,61 @@ public function isAbsolutePath($file)
     }
 
     /**
+     * Creates a temporary file with support for custom stream wrappers.
+     *
+     * @param string $dir    The directory where the temporary filename will be created.
+     * @param string $prefix The prefix of the generated temporary filename.
+     *                       Note: Windows uses only the first three characters of prefix.
+     *
+     * @return string The new temporary filename (with path), or FALSE on failure.
+     */
+    public function tempNam($dir, $prefix) {
+        $limit = 10;
+        $schemeSeparator = '://';
+        $components = explode($schemeSeparator, $dir, 2);
+        if (count($components) >= 2) {
+            $scheme = $components[0];
+            $dir = $components[1];
+        } else {
+            $scheme = false;
+            $dir = $components[0];
+        }
+        // Loop until we create a valid temp file or have reached $limit attempts
+        for ($i = 0; $i < $limit; $i++) {
+            // Create temp file in filesystem. We do this to use the file name generated by the OS.
+            $tmpFile = tempnam($dir, $prefix);
+            // If scheme is file
+            if ('file' === $scheme) {
+               // If the directory is the same
+                if (realpath($dir) === dirname($tmpFile)) {
+                    // Add the scheme back on
+                    $tmpFile = $scheme.$schemeSeparator.$dir.DIRECTORY_SEPARATOR.basename($tmpFile);
+                }
+            }
+            // If scheme exists and is valid
+            else if (false !== $scheme && file_exists($dir)) {
+                // Remove temp file in filesystem
+                unlink($tmpFile);
+                // Use the scheme version
+                $tmpFile = $scheme.$schemeSeparator.$dir.'/'.basename($tmpFile);
+                // If the file already exists set to FALSE and try again
+                if (file_exists($tmpFile)) {
+                    $tmpFile = false;
+                    continue;
+                }
+                // Create the file, if unsuccessful then stream is not writable so set to FALSE
+                // We use file_put_contents instead of touch as it support stream wrappers
+                if (false === @file_put_contents($tmpFile,'')) {
+                    $tmpFile = false;
+                }
+            }
+            break;
```

Comment:
Personally I find this method very confusing because of the way the code is structured. I would suggest to omit the break in the end and put it in a guard clause and try to avoid nesting if statements.

Please note the code style as well :)